### PR TITLE
Use KtFile.name instead of storing an absolute path key

### PR DESCRIPTION
**Pull request by arturbosch**

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt [hunk 1]
```diff
@@ -20,21 +22,43 @@ internal fun PsiElement.searchClass(): String {
...
+ * KtCompiler wrongly used Path.filename as the name for a KtFile instead of the whole path.
+ * This resulted into the question "How do we get the absolute path from a KtFile?".
+ * Fixing this problem, we do not need KtFile.absolutePath anymore.
+ *
+ * Fixing the filename will change all baseline signatures.
+ * Therefore we patch the signature here to restore the old behavior.
+ *
+ * Fixing the baseline will need a new major release - #2680.
+ */
 internal fun PsiElement.buildFullSignature(): String {
```
Comment (cortinico) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt:43]: Should we add a test for this extension function?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
 import org.jetbrains.kotlin.psi.KtFile
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.psiUtil.endOffset
 import org.jetbrains.kotlin.psi.psiUtil.getNonStrictParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
 import org.jetbrains.kotlin.psi.psiUtil.parents
 import org.jetbrains.kotlin.psi.psiUtil.startOffset
 
-private val signatureRegex = Regex("\\s(\\s|\t)+")
+private val multipleWhitespaces = Regex("\\s(\\s|\t)+")
```
Comment (cortinico) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt:15]:
```suggestion
private val multipleWhitespaces = Regex("\\s{2,}")
```
This should suffice and is also more aligned to the name of the variable.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt [hunk 3]
```diff
@@ -20,21 +22,43 @@ internal fun PsiElement.searchClass(): String {
     val classElement = this.getNonStrictParentOfType<KtClassOrObject>()
     var className = classElement?.name
     if (className != null && className == "Companion") {
-        classElement?.parent?.getNonStrictParentOfType<KtClassOrObject>()?.name?.let {
-            className = "$it.$className"
+        val parentClassName = classElement?.getStrictParentOfType<KtClassOrObject>()?.name
+        if (parentClassName != null) {
+            className = "$parentClassName.$className"
```
Comment (BraisGabin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt:27]: 🤔 couldn't this lead to things like `"asdf$null"`?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt (hunk 1 above)
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt:43]: This is tested through `Entity.from` in `EntitySpec` ^^

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt (hunk 3 above)
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/Signatures.kt:27]: This block is protected with a `class != null` if above :)

**Pull request merged**